### PR TITLE
Call only the User Actions api when the User action button is click SFINT-2544

### PR DESCRIPTION
**Pull request by jeremierobert-coveo**
# What was changed ?
+ Remove some caching here and there.
+ Added a model clear on user actions hide

All changes are covered by UT

### CODE REVIEW

**Comment by louis-bompart** [comment]
If there's an issue with file styling on the main branch, split that from your PR.

File: src/components/UserActions/ClickedDocumentList.ts [hunk 1]
```diff
@@ -75,9 +75,11 @@ export class ClickedDocumentList extends Component {
                 `<div class="coveo-list-row">
                     <div class="coveo-row-icon">${duplicate}</div>
                     <a class="CoveoResultLink"/a>
-                </div>`, {
```
Comment (louis-bompart) [inline src/components/UserActions/ClickedDocumentList.ts]: Out of scope of the PR imo

File: src/components/UserActions/QueryList.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ const DEFAULT_TRANSFORMATION = () => (query: string) => {
     const icon = document.createElement('div');
     icon.classList.add('coveo-row-icon');
     icon.innerHTML = search;
-
```
Comment (louis-bompart) [inline src/components/UserActions/QueryList.ts]: Out of scope of the PR imo

File: src/components/UserActions/UserActions.scss [hunk 3]
```diff
@@ -84,7 +84,6 @@ $search-ui-dropdown-header-color: #1d4f76;
         list-style-type: none;
 
         li {
-
```
Comment (louis-bompart) [inline src/components/UserActions/UserActions.scss]: Out of scope of the PR imo

File: src/components/UserActions/UserActions.scss [hunk 4]
```diff
@@ -96,7 +95,7 @@ $search-ui-dropdown-header-color: #1d4f76;
                 .coveo-row-icon {
                     position: relative;
                     width: $icon-container-size;
-
```
Comment (louis-bompart) [inline src/components/UserActions/UserActions.scss]: Out of scope of the PR imo

File: tests/components/UserActions/ClickedDocumentList.spec.ts [hunk 5]
```diff
@@ -115,10 +115,10 @@ describe('ClickedDocumentList', () => {
         return delay(() => {
             const list = mock.env.element.querySelector<HTMLOListElement>('.coveo-list');
 
-            for(let i=0; i<4; i++){
```
Comment (louis-bompart) [inline tests/components/UserActions/ClickedDocumentList.spec.ts]: Out of scope of the PR imo

File: tests/components/UserActions/ClickedDocumentList.spec.ts [hunk 6]
```diff
@@ -238,7 +238,7 @@ describe('ClickedDocumentList', () => {
         let getActionStub: SinonStub<[HTMLElement, ClickedDocumentList], void>;
         const mock = Mock.advancedComponentSetup<ClickedDocumentList>(
             ClickedDocumentList,
-            new Mock.AdvancedComponentSetupOptions(null, {userId: null}, env => {
```
Comment (louis-bompart) [inline tests/components/UserActions/ClickedDocumentList.spec.ts]: Out of scope of the PR imo

File: tests/components/UserActions/QueryList.spec.ts [hunk 7]
```diff
@@ -106,12 +106,12 @@ describe('QueryList', () => {
         return delay(() => {
             const list = mock.env.element.querySelector<HTMLOListElement>('.coveo-list');
 
-            for(let i=0; i<4; i++){
+            for (let i = 0; i < 4; i++) {
```
Comment (louis-bompart) [inline tests/components/UserActions/QueryList.spec.ts]: Out of scope of the PR imo

File: tests/components/UserActions/QueryList.spec.ts [hunk 8]
```diff
@@ -307,12 +307,12 @@ describe('QueryList', () => {
                 expect(mock.cmp.disabled).toBe(true);
             });
         });
-
```
Comment (louis-bompart) [inline tests/components/UserActions/QueryList.spec.ts]: Out of scope of the PR imo

File: tests/components/UserActions/QueryList.spec.ts [hunk 9]
```diff
@@ -307,12 +307,12 @@ describe('QueryList', () => {
                 expect(mock.cmp.disabled).toBe(true);
             });
         });
-    
+
         it('Should disable itself when the userId is empty string', () => {
             let getActionStub: SinonStub<[HTMLElement, QueryList], void>;
             const mock = Mock.advancedComponentSetup<QueryList>(
                 QueryList,
-                new Mock.AdvancedComponentSetupOptions(null, {userId: ''}, env => {
```
Comment (louis-bompart) [inline tests/components/UserActions/QueryList.spec.ts]: Out of scope of the PR imo

File: tests/components/UserActions/UserActivity.spec.ts [hunk 10]
```diff
@@ -532,7 +532,7 @@ describe('UserActivity', () => {
             let getActionStub: SinonStub<[HTMLElement, UserActivity], void>;
             const mock = Mock.advancedComponentSetup<UserActivity>(
                 UserActivity,
-                new Mock.AdvancedComponentSetupOptions(null, {userId: null}, env => {
+                new Mock.AdvancedComponentSetupOptions(null, { userId: null }, env => {
```
Comment (louis-bompart) [inline tests/components/UserActions/UserActivity.spec.ts]: Out of scope of the PR imo

File: tests/components/UserActions/UserActivity.spec.ts [hunk 11]
```diff
@@ -542,12 +542,12 @@ describe('UserActivity', () => {
                 expect(mock.cmp.disabled).toBe(true);
             });
         });
-
```
Comment (louis-bompart) [inline tests/components/UserActions/UserActivity.spec.ts]: Out of scope of the PR imo

File: tests/models/UserProfilingModel.spec.ts [hunk 12]
```diff
@@ -299,6 +299,41 @@ describe('UserProfilingModel', () => {
                 });
             });
 
+            it('should fetch all actions of a user from the backend only once', async () => {
```
Comment (louis-bompart) [inline tests/models/UserProfilingModel.spec.ts]: Your tests in this file are a bit complex, I think that's normal, however, I think we could improve them by adding some comments to makes them easier to understand at first glance ;)

Comment (jeremierobert-coveo) [reply tests/models/UserProfilingModel.spec.ts]: Read the it string :trollface: . (Yeah will add a little something)

File: src/models/UserProfileModel.ts [hunk 13]
```diff
@@ -90,16 +95,29 @@ export class UserProfileModel extends Model {
...
+    }
+
+    /**
+     * Delete all actions related to a user from the model.
+     * 
+     * @param userId The identifier of a user.
+     */
+    public deleteActions(userId: string) {
+        this.set(userId, undefined, UserProfileModel.MODEL_CONFIG);
+        this.getOrFetchCache[userId] = undefined;
```
Comment (Neilerino) [inline src/models/UserProfileModel.ts]: So to delete the actions here are we setting it as `undefined`?

File: src/rest/UserProfilingEndpoint.ts [hunk 14]
```diff
@@ -109,20 +104,8 @@ export class UserProfilingEndpoint {
...
-
-        if (!request) {
-            request = this.buildGetActionRequest(userId);
-            this.pendingGetActions[userId] = request;
-        }
-
-        return request;
-    }
-
-    private async buildGetActionRequest(userId: string) {
```
Comment (Neilerino) [inline src/rest/UserProfilingEndpoint.ts]: What's the reason for removing this functions functionality, and letting it skip straight to the `buildGetActionRequest` functionality?

File: src/models/UserProfileModel.ts (hunk 13 above)
Comment (jeremierobert-coveo) [reply src/models/UserProfileModel.ts]: Yess exactly 😄

File: src/rest/UserProfilingEndpoint.ts (hunk 14 above)
Comment (jeremierobert-coveo) [reply src/rest/UserProfilingEndpoint.ts]: It's one of the caching functionnality that I removed.

**Comment by louis-bompart** [comment]
@tedre191 can you change the threshold of Coveralls ?

**Comment by jeremierobert-coveo** [comment]
image

File: src/models/UserProfileModel.ts [hunk 15]
```diff
@@ -59,6 +59,11 @@ export class UserProfileModel extends Model {
         FETCH_CLICKED_DOCUMENT_FAIL: 'Fetching clicked documents details failed'
     });
 
+    private static readonly MODEL_CONFIG = {
+        customAttribute: true,
+        silent: true
```
Comment (tedre191) [inline src/models/UserProfileModel.ts]: Are you sure it needs to be silent ? Won't that prevent sending updates or something ?

Comment (jeremierobert-coveo) [reply src/models/UserProfileModel.ts]: That's what we want

**Pull request merged**